03_DeviceSelectionAndSharedSources/main.cpp

```diff
@@ -13,318 +13,422 @@ using namespace system;
-						for (auto j=0; j<binding.descriptorCount; j++)
-						{
-							auto& info = infos.emplace_back();
-							info.desc = buffers[j];
-							info.info.buffer = {0u,buffers[j]->getSize()};
-						}
-					}
-				}
-				// fix up the info pointers
-				for (auto& write : writes)
-					write.info = infos.data()+ptrdiff_t(write.info);
-				if (!m_device->updateDescriptorSets(writes.size(),writes.data(),0u,nullptr))
-					return logFail("Failed to write Descriptor Sets");
+				inputs[0][i] = i;
+				inputs[1][i] = DWORDCount - i;
 			}
-
-			// Make a utility since we have to touch 3 buffers
-			auto mapBuffer = [&]<typename PtrT>(const PtrT& buff, bitflag<IDeviceMemoryAllocation::E_MAPPING_CPU_ACCESS_FLAGS> accessHint) -> uint32_t*
+			for (auto j = 0; j < 2; j++)
 			{
-				auto* const memory = buff->getBoundMemory();
-				const IDeviceMemoryAllocation::MappedMemoryRange memoryRange(memory,0ull,memory->getAllocationSize());
-
-				void* ptr = m_device->mapMemory(memoryRange,accessHint);
-				if (!ptr)
-					m_logger->log("Failed to map the whole Memory Allocation of a GPU Buffer %p for access %d!\n",ILogger::ELL_ERROR,buff,accessHint);
-				
-				if (accessHint.hasFlags(IDeviceMemoryAllocation::EMCAF_READ))
-				if (!memory->getMemoryPropertyFlags().hasFlags(IDeviceMemoryAllocation::EMPF_HOST_COHERENT_BIT))
-					m_device->invalidateMappedMemoryRanges(1,&memoryRange);
-
-				return reinterpret_cast<uint32_t*>(ptr);
+				const auto& buffMem = inputBuff[j].second.memory;
+				// New thing to learn, if the mapping is not coherent and you write it, you need to flush!
+				const ILogicalDevice::MappedMemoryRange memoryRange(buffMem.get(), 0ull, buffMem->getAllocationSize());
+				m_device->flushMappedMemoryRanges(1, &memoryRange);
+				buffMem->unmap();
+			}
+		}
+
+#ifndef USE_INTROSPECTOR
+		{
+			IGPUDescriptorSet::SDescriptorInfo inputBuffersInfo[2];
+			inputBuffersInfo[0].desc = smart_refctd_ptr(inputBuff[0].first);
+			inputBuffersInfo[0].info.buffer = { .offset = 0,.size = inputBuff[0].first->getSize() };
+			inputBuffersInfo[1].desc = smart_refctd_ptr(inputBuff[1].first);
+			inputBuffersInfo[1].info.buffer = { .offset = 0,.size = inputBuff[1].first->getSize() };
+
+			IGPUDescriptorSet::SDescriptorInfo outputBufferInfo;
+			outputBufferInfo.desc = smart_refctd_ptr(outputBuff.first);
+			outputBufferInfo.info.buffer = { .offset = 0,.size = outputBuff.first->getSize() };
+
+			IGPUDescriptorSet::SWriteDescriptorSet writes[2] = {
+				{.dstSet = ds[1].get(),.binding = 2,.arrayElement = 0,.count = 2,.info = inputBuffersInfo},
+				{.dstSet = ds[3].get(),.binding = 6,.arrayElement = 0,.count = 1,.info = &outputBufferInfo}
 			};
-			
-			// We'll fill the buffers in a way they always add up to WorkgrouCount*WorkgroupSize
-			constexpr auto DWORDCount = WorkgroupSize*WorkgroupCount;
+
+			m_device->updateDescriptorSets(std::span(writes, 2), {});
+		}
+#endif
```

Comment:
would be cool if you could use the Stage Introspection data to find out the values you need to populate `SWriteDescriptorSet` with (binding and set).

But I guess that would require you to "know" the names of the descriptor variables of the in the stage introspection.